### PR TITLE
Snyder - LB141.ag_Fert_IFA_ctry_crop

### PR DESCRIPTION
**Pull request by abigailsnyder**
Review from @pkyle especially focusing on comments would be appreciated.

The `AGLU_ctry` handling I brought up in https://github.com/JGCRI/gcamdata/issues/364 is covered in the pair of pipelines starting on Line 171. It’s probably not the best way to handle it and I’m open to suggestions for improvements. I did fill in some data in `AGLU_ctry` to make the code a little nicer:
Fille in IFA2002_country = CROATIA for Yugoslav FSR + iso = hrv. 
The other Yugoslav FSR + iso combos also are missing IFA2002_country info. I checked if any of the individual countries/iso options have their own IFA2002_country info to fill in, but none of the others did.
In other words: Bosnia/Herz.; Macedonia; Montserrat; Serbia; and Slovenia do not have IFA2002_country info. 
I added corresponding IFA_region info to Yugoslav/iso for any iso that had it (Serbia and Montserrat do not, I believe). I don’t know if this missing data is an issue or not. 

Also still waiting on information from https://github.com/JGCRI/gcamdata/issues/364 with respect to a possibly inaccurate comment and/or code chunk on lines 65-68 in old DS. PR does currently match old DS and pass tests, and if any changes need to be made, they should be in the `OLD_DATA_SYSTEM_BEHAVIOR` format. So this portion can still be reviewed, I believe.

### CODE REVIEW

File: R/constants.R [hunk 1]
```diff
@@ -98,6 +99,8 @@ CONV_THA_KGM2 <- 0.1   # tons C/ha -> kg C/m2
 CONV_GG_TG <- 0.001 # gigagrams to tegagrams
 CONV_TST_TG <- 0.000907 # thousand short tons to Tg
 CONV_KG_TO_TG <- 1e-9
+CONV_KT_MT <- 0.001 #kt to Mt
```
Comment (bpbond) [inline R/constants.R]: Space after `#` pls

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 2]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+    # Lines 65-69 in original file
+    # Convert IFA fertilizer consumption to long form and convert units to Mt of N consumed.
+    # This is top down fertilizer consumption data, to be reconciled with bottom up consumption
+    # estimate using IFA2002 fertilizer consumption, and LDS, FAO and Monfreda harvested area data.
+    IFA_Fert_ktN %>%
+      filter(Country != "World Total") %>%
+      gather(IFA_commodity, value, -Country) %>%
+      # convert units from kt to Mt of Nitrogen
+      mutate(Fert_MtN = value * CONV_KT_MT) %>%
+      # rename Country
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: For what it's worth, I personally find this kind of comment not helpful; it takes up screen real estate, and doesn't tell us anything more than a glance at the code would. You can and should assume basic (programming) literacy on the part of the reader, I think.

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 3]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+      rename(IFA_region = Country) %>%
+      # rename value to be more informative
+      rename (Fert_ktN = value) ->
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: Could combined into a single `rename` call

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 4]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+      filter(year %in% FAO_LDS_YEARS) %>%
+      group_by(GCAM_region_ID, GCAM_commodity) %>%
+      summarise(FAO = mean(value)) %>%
+      ungroup() ->
+      L141.FAO
+    # Take LDS HA data and aggregate to the GCAM region-commodity level:
+    L102.ag_HA_bm2_R_C_GLU %>%
+      group_by(GCAM_region_ID, GCAM_commodity) %>%
+      summarise(LDS = sum(value)) %>% ungroup() %>%
+      ungroup() %>%
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: Two ungroups in a row

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 5]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+      # Join in the FAO data from the previous pipeline
+      left_join_error_no_match(L141.FAO, by = c("GCAM_region_ID", "GCAM_commodity")) %>%
+      # Calculate the FAO_LDS scaler value = FAO/LDS
+      mutate(scaler = FAO/LDS) ->
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: Spaces around `/` pls

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 6]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+      summarise(FAO = mean(value)) %>%
+      ungroup() ->
+      L141.FAO
+    # Take LDS HA data and aggregate to the GCAM region-commodity level:
+    L102.ag_HA_bm2_R_C_GLU %>%
+      group_by(GCAM_region_ID, GCAM_commodity) %>%
+      summarise(LDS = sum(value)) %>% ungroup() %>%
+      ungroup() %>%
+      # Join in the FAO data from the previous pipeline
+      left_join_error_no_match(L141.FAO, by = c("GCAM_region_ID", "GCAM_commodity")) %>%
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: Perhaps remove the `ungroup` calls above and just do it once, here?

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 7]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+
+    # Second, use the region-commodity scalers to scale LDS harvested area data at the iso-GLU-GTAPcrop level.
+    # old comment: Match the GCAM region and commodity class into the LDS dataset, and multiply by the scaler above
+    #
+    # Take LDS harvested area data at the iso-GLU-GTAPcrop level, L100.LDS_ag_HA_ha, join in the GCAM_region_ID info for
+    # each iso, join in the GCAM_commodity info for each GTAPcrop, join the FAO_LDS scaler calculated in the previous
+    # pipeline, and use the FAO_LDS scaler to scale the iso-GLU-GTAPcrop harvested area data
+    L100.LDS_ag_HA_ha %>%
+      # join GCAM region ID information
+      left_join_error_no_match(select(iso_GCAM_regID, iso, GCAM_region_ID), by = c("iso")) %>%
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: `by = "iso"` (same comment applies below; no need for `c()` when only a single item)

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 8]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+      # join GCAM commodity information and preserve NAs to be omitted, as in original code
+      left_join(select(FAO_ag_items_PRODSTAT, GTAP_crop, GCAM_commodity), by = c("GTAP_crop")) %>%
+      na.omit() %>%
+      # join FAO_LDS scaler for each region-commodity
+      left_join_error_no_match(select(L141.FAO_LDS, GCAM_region_ID, GCAM_commodity, scaler),
+                               by = c("GCAM_region_ID", "GCAM_commodity")) %>%
+      # use the region-commodity scaler to scale down each iso-GLU-GTAP harvested area
+      mutate(value = value * scaler) %>%
+      # drop unwanted columns
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: Again, not super helpful. This is only my opinion; feel free to leave if you find useful.

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 9]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+    # This ultimately leads to the bottom-up estimate of fertilizer consumption at the iso-GTAPcrop level, using scaled
+    # LDS data from L141.LDS_ag_HA_ha, IFA2002 fertlizer consumption data, and the top down consumption data in L141.IFA_Fert_ktN.
+    #
+    # Calculate a bottom up consumption RATE of N in units of tons/hectare for each IFA2002 country and crop.
+    # There is a time component of IFA2002 that is handeled by taking the max values over time for each country-crop.
+    # specifically, the rate is calculated first and THEN the max is taken, rather than taking the rate as
+    # max N_kt / max Area_ha,which has the possibility of cancellation.
+    IFA2002_Fert_ktN %>%
+      # calculate nitrogen consumption in tons per hectare N_tha = N_kt / AREA_thousHa
+      mutate(N_tha = N_kt /AREA_thousHa) %>%
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: Nice naming 👍

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 10]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+    # Some IFA_region information was added to other Yugoslav/iso's when present for corresponding iso's (to allow
+    # the call to unique).
+    # This avoids the use of left_join_keep_first_only in the subsequent pipeline.
+    AGLU_ctry %>%
+      # take only the isos we need
+      filter(iso %in% L141.LDS_ag_HA_ha$iso) %>%
+      # keep only the columns we need
+      select(iso, IFA2002_country, IFA_region) %>%
+      # keep unique
+      unique() ->
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: `dplyr::distinct()` is a lot faster than `unique`

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 11]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+    #
+    # Pipeline 2: Use this scaler to scale the bottom-up estimates of fertilizer consumption at the iso-GTAPcrop level in
+    # table L141.IFA_Fert_Cons_MtN_ctry_crop
+
+    # Pipeline 1:
+    # take ctry crop bottom up estimate of N consumption
+    L141.IFA_Fert_Cons_MtN_ctry_crop %>%
+      # keep relevant columns only
+      select(IFA_region, IFA_commodity, IFA_N_Mt_unscaled) %>%
+      # aggregte to IFA_region, IFA_commodity via summing
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: "aggregate"

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 12]
```diff
@@ -40,43 +42,259 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+      group_by(IFA_region, IFA_commodity) %>%
+      summarise(IFA_N_Mt_unscaled = sum(IFA_N_Mt_unscaled)) %>%
+      ungroup() %>%
+      # join top down IFA fertilizer estimates
+      left_join_error_no_match(select(L141.IFA_Fert_ktN, -Fert_ktN), by = c("IFA_region", "IFA_commodity")) %>%
+      # calculate scaler = top down / bottom up N consumption = Fert_MtN / IFA_N_Mt_unscaled
+      mutate(scaler = Fert_MtN / IFA_N_Mt_unscaled) %>%
+      # replace any NA scalers witha  value of 1
```
Comment (bpbond) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: "with a"

File: inst/extdata/aglu/AGLU_ctry.csv [hunk 13]
```diff
@@ -1 +1,277 @@
...
Yugoslav SFR,bih,xer,,,,,,12,Central Europe,Transition countries
Yugoslav SFR,hrv,xer,,,,,,12,Central Europe,Transition countries
Yugoslav SFR,mkd,xer,,,,,,12,Central Europe,Transition countries
Yugoslav SFR,mne,xer,,,,,,12,Central Europe,Transition countries
Yugoslav SFR,srb,xer,,,,,,12,Central Europe,Transition countries
Yugoslav SFR,svn,xer,,,,,,12,Central Europe,Transition countries
Zambia,zmb,zmb,ZAMB,37,Zambia,ZAMBIA,Rest of world,10,Southern Africa,sub-Saharan Africa
Zimbabwe,zwe,zwe,ZIMB,32,Zimbabwe,ZIMBABWE,Rest of world,10,Southern Africa,sub-Saharan Africa
\ No newline at end of file
+# File: AGLU_ctry.csv
```
Comment (bpbond) [inline inst/extdata/aglu/AGLU_ctry.csv]: Thanks for fixing this file!

**Comment by bpbond** [comment]
> I did fill in some data in AGLU_ctry to make the code a little nicer

Hmm. We'll have to make sure no one overwrites those changes in a future PR then, right?

**Comment by abigailsnyder** [comment]
@bpbond  in theory yes. But it was strictly filling in missing data for Yugoslavia. If I'm interpreting what Page said in https://github.com/JGCRI/gcamdata/issues/364 correctly (no guarantees), countries like Yugoslavia that dissolved have just not had their entries completely updated in `AGLU_ctry`. So there _shouldn't_ be any instance where someone would need to overwrite what I filled in to something different

**Comment by pkyle** [comment]
@bpbond just for clarification, the column that @abigailsnyder was working on in AGLU_ctry is not used by any other code chunks.

**Comment by pkyle** [comment]
@abigailsnyder Hope you don't mind, I just pushed a commit with changes to comments. Didn't change any code (re-tested just to make sure). I removed all references to line #s in the old data system because I figured we won't need 'em anymore...feel free to put back of course.

**Comment by abigailsnyder** [comment]
@pkyle, I super appreciate the extra comments. 
I've been including old line numbers but I can totally stop doing that in general in my code.

FYI re `ALGU_ctry`, I ended up tweaking the Yugo-Croatia entry for `IFA2002_country`, AND a couple of the Yugo `IFA_region` entries (making sure they always correspond with the individual country's `IFA_region`). This let me do a single pipeline to clean up what we need from `AGLU_ctry` and only join from it once, instead of in two places in the original code.

File: R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R [hunk 14]
```diff
@@ -40,43 +42,237 @@ module_aglu_LB141.ag_Fert_IFA_ctry_crop_DISABLED <- function(command, ...) {
...
+
+    # Use the bottom-up country crop table of harvested area and N application rate, L141.IFA_Fert_Cons_MtN_ctry_crop,
+    # to calculate specific bottom up fertilizer demand = harvested area * application rate.
+    # When such information is missing from the bottom up tables, use the default estimate from the top-down estimate
+    # in table L141.HA_ha_Rifa_Cifa.
+    L141.IFA_Fert_Cons_MtN_ctry_crop %>%
+      # join in relevent IFA_region, IFA_commodity information
+      left_join_error_no_match(select(L141.HA_ha_Rifa_Cifa, IFA_region, IFA_commodity, IFA_N_tha_default),
+                               by = c("IFA_region", "IFA_commodity")) %>%
+      # calculate bottom up fertilizer demand = harvested area * consumpstion rate, using defaults where missing
```
Comment (kvcalvin) [inline R/zchunk_LB141.ag_Fert_IFA_ctry_crop.R]: "consumption"
